mutation_partition_v2.cc

```diff
@@ -0,0 +1,1154 @@
+/*
+ * Copyright (C) 2014-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+#include <boost/range/adaptor/reversed.hpp>
+#include "mutation_partition_v2.hh"
+#include "clustering_interval_set.hh"
+#include "converting_mutation_partition_applier.hh"
+#include "partition_builder.hh"
+#include "query-result-writer.hh"
+#include "atomic_cell_hash.hh"
+#include "reversibly_mergeable.hh"
+#include "mutation_fragment.hh"
+#include "mutation_query.hh"
+#include "service/priority_manager.hh"
+#include "mutation_compactor.hh"
+#include "counters.hh"
+#include "row_cache.hh"
+#include "view_info.hh"
+#include "mutation_cleaner.hh"
+#include <seastar/core/execution_stage.hh>
+#include "types/map.hh"
+#include "compaction/compaction_garbage_collector.hh"
+#include "utils/exceptions.hh"
+#include "clustering_key_filter.hh"
+#include "mutation_partition_view.hh"
+#include "tombstone_gc.hh"
+
+extern logging::logger mplog;
+
+template<bool reversed>
+struct reversal_traits;
+
+template<>
+struct reversal_traits<false> {
```

Comment:
On the C++ standard level, it's a part of the One Definition Rule (https://en.cppreference.com/w/cpp/language/definition#One_Definition_Rule), which (among other things), says that all definitions in the program of the same external-linkage name have to be exactly the same (up to whitespace), or the program is ill-formed.

On the practical/toolchain level, if you have two definitions in the program, the linker will pick one arbitrarily (actually: the first one it encounters) and it will be used for *all* references in the program. Definitions in the same file don't get "priority". So if these duplicate classes/functions diverge, one of them will be unused and one will be used by both files, which can lead to confusing bugs. (Made even more confusing due to the fact that *both* versions can be used together in case of inlining — inlined calls will use the *local* definition, non-inlined calls will use the arbitrarily-picked by the linker *global* definition.

Reply:
So if some unrelated part of code comes up with the same name for a template by some chance, C++ will silently generate a corrupted program? What a great language.

Reply:
I know, right?

Reply:
Well, it's one reason to make lots of namespaces. The namespace is a part of the name, so they make it harder for unrelated parts of the program to interfere with each other at link time.

Reply:
Oh, I forgot to mention that you can use unnamed namespaces to force internal linkage. (Unlike `static` this works not only on variables and functions, but also on classes). But that's easy to mess up, (e.g. you won't see in a review diff that someone moved a class out of a unnamed namespace...), so I think it's better to just avoid name conflicts, rather than patch them with namespaces.